### PR TITLE
Use Task class instead of tuple

### PR DESCRIPTION
**Pull request by fjetter**
This is an early version that will close https://github.com/dask/dask/issues/9969

It introduces a new `Task` class (name is subject to change) and a couple of other related subclasses that should replace the tuple as a representation of runnable tasks.

The benefits of this are outlined in https://github.com/dask/dask/issues/9969 but are primarily focused to reduce overhead during serialization and parsing of results. An important result is also that we can trivially cache functions (and arguments if we wish) to avoid problems like https://github.com/dask/distributed/issues/8767 where users are erroneously providing expensive to pickle functions (which also happens frequently in our own code and/or downstream projects like xarray)

This approach allows us to convert the legacy dsk graph to the new representation with full backwards compatibility. Old graphs can be migrated and new ones written directly using this new representation which will ultimately reduce overhead.

I will follow up with measurements shortly.

Sibling PR in dask https://github.com/dask/dask/pull/11248

### CODE REVIEW

File: distributed/client.py [hunk 1]
```diff
@@ -939,10 +939,12 @@ def _construct_graph(self) -> _T_LowLevelGraph:
                     dsk.update(vv.dask)
                 else:
                     kwargs2[k] = v
+                from dask.task_spec import Task
 
                 dsk.update(
                     {
-                        key: (apply, self.func, (tuple, list(args)), kwargs2)
+                        key: Task(key, self.func, args, kwargs2)
+                        # (apply, self.func, (tuple, list(args)), kwargs2)
```
Comment (fjetter) [inline distributed/client.py:947]: This isn't necessary but an example of how this conversion would look like

File: distributed/scheduler.py [hunk 2]
```diff
@@ -773,17 +772,20 @@ def remove_from_processing(self, ts: TaskState) -> None:
...
+            del tp_count[prefix_name]
 
-        count = self.scheduler._task_prefix_count_global[ts.prefix.name] - 1
+        count = tp_count_global[prefix_name] - 1
         if count:
-            self.scheduler._task_prefix_count_global[ts.prefix.name] = count
+            tp_count_global[prefix_name] = count
         else:
-            del self.scheduler._task_prefix_count_global[ts.prefix.name]
+            del tp_count_global[prefix_name]
```
Comment (fjetter) [inline distributed/scheduler.py:788]: this is an unrelated perf fix

File: distributed/scheduler.py [hunk 3]
```diff
@@ -9062,34 +9068,8 @@ def _materialize_graph(
...
-        deps = set(v)
-        deps.discard(k)
-        dependencies[k] = deps
-
-    dsk = valmap(_normalize_task, dsk)
+    dsk = convert_old_style_dsk(dsk)
+    # TODO: This isn't working yet as expected
+    dependencies = dict(DependenciesMapping(dsk))
+
     return dsk, dependencies, annotations_by_type
```
Comment (fjetter) [inline distributed/scheduler.py:9075]: most/all of this complexity is now either gone entirely or hidden in the class

**Review by jacobtomlinson — COMMENTED** [review]
There may be implications for some of the dashboard components, the "pew pew pew" plot comes to mind. I see this is still a draft, let me know when it's in a reviewable state and I'll look over the dashboard code to see if anything needs changing there 🙂.

**Comment by fjetter** [comment]
> There may be implications for some of the dashboard components, the "pew pew pew" plot comes to mind. I see this is still a draft, let me know when it's in a reviewable state and I'll look over the dashboard code to see if anything needs changing there 🙂.

I'd actually be surprised if that was affected since we don't change the scheduler internal metadata (like dependencies, transfers, where the tasks are executed...). But who knows. I'll probably stumble over 50 small weird things trying to get CI green :)

File: distributed/scheduler.py (hunk 2 above)
Comment (hendrikmakait) [reply distributed/scheduler.py:790]: I'm curious, how noticeable is this? Apart from that, let's move this to a separate PR to keep this focused on the major change you introduce here.

File: distributed/scheduler.py [hunk 4]
```diff
@@ -9046,11 +9055,12 @@ def transition(
 
 
 def _materialize_graph(
-    graph: HighLevelGraph, global_annotations: dict[str, Any]
+    graph: HighLevelGraph, global_annotations: dict[str, Any], validate: bool
 ) -> tuple[dict[Key, T_runspec], dict[Key, set[Key]], dict[str, dict[Key, Any]]]:
     dsk = ensure_dict(graph)
-    for k in dsk:
-        validate_key(k)
+    if validate:
```
Comment (hendrikmakait) [inline distributed/scheduler.py:9061]: There are a few seemingly unrelated changes in this file. Let's extract them to a separate PR.

File: distributed/scheduler.py (hunk 2 above)
Comment (fjetter) [reply distributed/scheduler.py:790]: https://github.com/dask/distributed/pull/8821

File: distributed/scheduler.py (hunk 4 above)
Comment (fjetter) [reply distributed/scheduler.py:9363]: https://github.com/dask/distributed/pull/8822

File: distributed/scheduler.py [hunk 5]
```diff
@@ -4907,7 +4917,31 @@ async def update_graph(
...
+                "status": "OK",
+            }
+            self.log_event(["all", client, "update-graph"], evt_msg)
+            logger.debug("Task state created. %i new tasks", len(self.tasks) - before)
+        except Exception as e:
+            evt_msg = {
+                "action": "update-graph",
+                "stimulus_id": stimulus_id,
+                "status": "error",
+            }
```
Comment (fjetter) [inline distributed/scheduler.py:4943]: This is an unrelated change but it shouldn't be too disruptive for the review

File: distributed/scheduler.py [hunk 6]
```diff
@@ -4872,6 +4872,13 @@ async def update_graph(
                 global_annotations=annotations or {},
                 validate=self.validate,
             )
+
+            materialization_done = time()
+            logger.debug("Materialization done. Got %i tasks.", len(dsk))
+            dependents = reverse_dict(dependencies)
+            dsk = resolve_aliases(dsk, keys, dependents)
+            dependencies = dict(DependenciesMapping(dsk))
+            logger.debug("Removing aliases. %i left", len(dsk))
```
Comment (fjetter) [inline distributed/scheduler.py:4881]: This is new... it removes all those 

`"key": "other-key"`

references that we currently schedule as real tasks. Particularly fusion adds these kinds of redirects. I should probably factor this out to a dedicated PR. The impact can be quite substantial in graph size reduction.

File: distributed/scheduler.py [hunk 7]
```diff
@@ -9360,33 +9399,18 @@ def _materialize_graph(
...
+    # This also performs culling!
+    dsk3 = resolve_aliases(dsk2, keys, dependents)
+
+    logger.debug(
+        "Removing aliases. Started with %i and got %i left", len(dsk2), len(dsk3)
+    )
+    # FIXME: There should be no need to fully materialize and copy this but some
+    # sections in the scheduler are mutating it.
+    dependencies = {k: set(v) for k, v in DependenciesMapping(dsk3).items()}
+    return dsk3, dependencies, annotations_by_type
```
Comment (fjetter) [inline distributed/scheduler.py:9416]: see https://github.com/dask/distributed/pull/8842

**Review by fjetter — COMMENTED** [review]
I recommend reviewers to start with the dask/dask PR.

This PR does not include (m)any intentional changes other than adjusting the code to the new spec. There are one or two things that change behavior, I flagged them explicitly

File: distributed/tests/test_client.py [hunk 8]
```diff
@@ -8402,6 +8237,9 @@ async def test_release_persisted_collection(c, s, a, b):
         await c.compute(arr)
 
 
+@pytest.mark.skip(
+    reason="Deadlocks likely related to future serialization and ref counting"
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/tests/test_client.py:8241]: that should've been fixed by https://github.com/dask/distributed/pull/8827

I'll remove the skip

File: distributed/scheduler.py [hunk 9]
```diff
@@ -4907,7 +4914,31 @@ async def update_graph(
                 start=start,
                 stimulus_id=stimulus_id or f"update-graph-{start}",
             )
-        except RuntimeError as e:
+            task_state_created = time()
+            metrics.update(
+                {
+                    "start": start,
+                    "duration_materialization": materialization_done - start,
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/scheduler.py:4921]: can you suggest the appropriate names? I'm not sure what the correct way is

File: distributed/scheduler.py [hunk 10]
```diff
@@ -4907,7 +4914,31 @@ async def update_graph(
...
+            }
+            self.log_event(["all", client, "update-graph"], evt_msg)
+            logger.debug("Task state created. %i new tasks", len(self.tasks) - before)
+        except Exception as e:
+            evt_msg = {
+                "action": "update-graph",
+                "stimulus_id": stimulus_id,
+                "status": "error",
+            }
+            self.log_event(["all", client, "update-graph"], evt_msg)
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/scheduler.py:4941]: No particular reason. I used this in a test but I will change it. I will also drop the `all` topic (feels a bit like an anti pattern)

File: distributed/scheduler.py (hunk 5 above)
Comment (fjetter) [reply distributed/scheduler.py:4940]: I don't think this is very useful and I don't want to add the exception to this. This is primarily supposed to be a stream of metrics and I don't like adding Exception objects or large strings to it. I also find the logging and proper exception handling mechanism should be sufficient.

If this is a contentious topic I will remove this change from the PR

File: distributed/tests/test_client.py [hunk 11]
```diff
@@ -4724,97 +4728,6 @@ def f(i, L=None):
     assert results == [i + 6 for i in range(10)]
 
 
-@gen_cluster(client=True)
-async def test_recreate_error_delayed(c, s, a, b):
```
[the remark this answers is not in the document]
Comment (hendrikmakait) [reply distributed/tests/test_client.py:4728]: Sounds good!

File: distributed/scheduler.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (hendrikmakait) [reply distributed/scheduler.py]: Note that dropping the `all` pattern is a user-facing breaking change. That being said, I'm all for redesigning our topics, etc., this might just require some changes for downstream users, e.g., for Coiled.

[the remark this answers is not in the document]
Comment (hendrikmakait) [reply distributed/scheduler.py]: Then again, we've also renamed the action, so these changes are breaking already.

[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/scheduler.py]: Yeah, I think this is breaking and I'm fine with it. This is overall a pretty burried feature and I doubt (m)any users will notice.

**Comment by fjetter** [comment]
I think this is slowly coming together.

I see some kind of unexpected failures still in `distributed/shuffle/tests/test_merge.py::test_merge`. They indicate some shuffle failed errors that I cannot see on any test report.
The other failures (so far) appear to be unrelated

**Comment by fjetter** [comment]
Test failures are all over the place. I don't think anything is related.

**Comment by fjetter** [comment]
I triggered another A/B just to make sure

**Comment by fjetter** [comment]
https://github.com/coiled/benchmarks/actions/runs/10679723905

The results of the A/B run are interesting. I will copy the test cases with some signal here

There are a couple of notable improvements already. This shows test_swap_axes_in_memory which is using P2P rechunking. P2P rechunking is already using the new task spec so the parsing is much cheaper. We're up to 25% faster

image

However, there are also a couple of cases for P2P to be now slower by 5-10%. For example

image
image

There are also some notable regression around Client.map (already using the new spec)

image

and some other array code (test_doublediff)

image

There are some notable gains with test_filter_then_average by up to 15% but these tests have been known to be a little flaky

image


So overall a bit of a mixed bag. I'm not very discouraged considering that we haven't migrated any other low level code to this system but I will poke at the performance regressions a bit to see if there are some easy wins.

**Comment by fjetter** [comment]
The Client.map perf regression was trivial and should be fixed by now. I only migrated the complex version with kwargs but not the case without. Therefore, we went from "basically doing nothing" to "let's parse this thing once" which was showing up on those tests that basically measured overhead.

**Comment by fjetter** [comment]
I'll have to look into this a little more, I think

I repeated the benchmarks and got

## Wall time

image

## Peak Memory (W) 

image


The only thing I'm not surprised about is that Scheduler Memory also increases (which is expected at this stage)

image

**Comment by fjetter** [comment]
I suspect that the memory increase, particularly for the P2P cases, is caused by the cache building up but I haven't checked. yet. In absolute values, the increase is much less severe as the above is letting believe

**Comment by fjetter** [comment]
So I briefly looked into `test_tiles_to_rows` (that one with the really bad memory result. I ran CPU and memory profiles but cannot find anything that's wrong with it.
The caches are populated but hold less than 10MB. The memray profile doesn't show anything out of the ordinary. The accumulated memory are the shards we're shuffling and not anything related to the spec itself.

I'll check again what happens if I disable the caching entirely but I can't imagine how this would change anything about the lifetyime of the shards.

**Comment by fjetter** [comment]
I changed CI configuration to run against main again. https://github.com/dask/dask/pull/11248 should be fine to merge and afterwards we'll retrigger CI

**Comment by fjetter** [comment]
needs https://github.com/dask/dask/pull/11429

**Comment by fjetter** [comment]
Also this https://github.com/dask/dask/pull/11431

**Comment by fjetter** [comment]
The mindeps builds are sad. Everything else seems unrelated

File: distributed/scheduler.py [hunk 12]
```diff
@@ -4609,27 +4614,38 @@ async def add_nanny(self, comm: Comm, address: str) -> None:
...
+                if d not in dsk:
+                    if d not in self.tasks:
+                        lost_keys.add(d)
+                        lost_keys.add(k)
+                        logger.info("User asked for computation on lost data, %s", k)
+                        dependencies.pop(d, None)
+                        keys.discard(k)
+                    continue
+                wupdate(dsk[d].dependencies)
+        return lost_keys
```
Comment (fjetter) [inline distributed/scheduler.py:4644]: I rewrote this section. I had issues with it and just barely understand the old code (and ran into multiple bugs in the recent past)

**Comment by hendrikmakait** [comment]
Update: `test_merge` failures seem systemic (and thus related).

File: distributed/scheduler.py [hunk 13]
```diff
@@ -4609,27 +4614,38 @@ async def add_nanny(self, comm: Comm, address: str) -> None:
...
+
+    def _remove_done_tasks(
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/scheduler.py:4646]: ah, likely a merge conflict since I renamed the method. this is indeed dead

**Comment by fjetter** [comment]
> Update: test_merge failures seem systemic (and thus related).

I've been trying to reproduce but without any luck so far

**Comment by fjetter** [comment]
I _think_ the test_merge failures are actually unrelated. The exception is

```python-traceback
  
        def validate_data(self, data: pd.DataFrame) -> None:
>           if set(data.columns) != set(self.meta.columns):
E           AttributeError: 'tuple' object has no attribute 'columns'
```

which indicates that `data` is likely a `key` instead of the data... I've seen this before and I think this is a dask-expr problem. I added another verification step here to confirm this. I'll keep digging

**Comment by fjetter** [comment]
yeah, so the exception is pretty much what I expected

```python-traceback
            if not isinstance(data, pd.DataFrame):
>               raise TypeError(f"Expected {data=} to be a DataFrame, got {type(data)}.")
E               TypeError: Expected data=('assign-3d7cfa7cea412465799bea6cfac1b512', 1) to be a DataFrame, got <class 'tuple'>.
```

**Comment by fjetter** [comment]
Ah, this is the build with dask-expr *disabled*. Now I can reproduce!

File: distributed/shuffle/_shuffle.py [hunk 14]
```diff
@@ -288,10 +289,10 @@ def _construct_graph(self) -> _T_LowLevelGraph:
         name = "shuffle-transfer-" + token
         transfer_keys = list()
         for i in range(self.npartitions_input):
-            transfer_keys.append((name, i))
-            dsk[(name, i)] = (
+            t = Task(
+                (name, i),
                 shuffle_transfer,
-                (self.name_input, i),
+                TaskRef((self.name_input, i)),
```
Comment (fjetter) [inline distributed/shuffle/_shuffle.py:295]: The `test_merge` tests w/ dask-expr enabled never take this code path. That's interesting but not incredibly surprising.

**Comment by fjetter** [comment]
`distributed/shuffle/tests/test_graph.py::test_multiple_linear` failure is also related. also a legacy-only problem

**Comment by fjetter** [comment]
https://github.com/dask/dask/pull/11445/ is hopefully the last one

**Comment by fjetter** [comment]
Sample size: Exactly one

I ran this on a large, sluggish workload that took ~20min to send/materialize/order/etc.
(It's actually 500k tasks, the upper left number is a bug)
image

With this PR this is cut in half

image

CPU load for the remainder of the run also seems to be significantly reduced but I haven't let the entire thing finish.

I noticed a significant memory spike that I still want to investigate before merging

**Comment by hendrikmakait** [comment]
Is there anything left to do here?
